SoftKiller

emadmtr opened the pull request:
This is my initial implementation of SoftKiller, I've created 4 files:
           - SoftKiller.jl - handles the logic behind the SoftKiller algorithm and the setup of the rectangular grid 
           - TilingBase.jl - keeps track of information about each tile in the grid
           - softkl.jl - example file of working softkiller with plots for comparison 
           - softly_rt.jl - example file without plotting, e.g. for time comparison between Julia and C++ implementations

This PR closes #123 


We are looking forward to your review and merging this in.  Thank you for your patience.

mattleblanc commented:
Hi @graeme-a-stewart, we weren't sure what to do with the formatting in the CI. When we ran `JuliaFormatter` it seemed to make changes that didn't cause that check to pass on my own fork, so we've left it undone for now.

We're anyway expecting to make some changes during review, so we thought this could be handled in parallel.

m-fila commented on src/JetReconstruction.jl:
```diff
@@ -67,6 +68,14 @@ export tiled_jet_reconstruct
+
+## SoftKiller
+include("SoftKiller.jl")
+export SoftKiller, apply, select_ABS_RAP_max, push_data, plot_set_up
```
```suggestion
export SoftKiller, apply, select_ABS_RAP_max
```
,`push_data` and `plot_set_up` are defined in examples and don't exist in the module

emadmtr replied:
I went with softkiller_apply but I can change it to just softkiller if you believe it could be more sufficient.

Moelf commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+    _nphi::Int64
+
+    function SoftKiller(rapmin::Float64, rapmax::Float64, drap::Float64, dphi::Float64)
+        print("4 variables \n")
```
are these debug?

graeme-a-stewart replied:
Should be removed

m-fila commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+    descr = "rectangular grid with rapidity extent $(sk._ymin) < rap < $(sk._ymax) \n total tiles  $(sk._ntotal) \n "
+    descr *= "tile size drap x dphi = $(sk._dy) x $(sk._dphi)"
+
+    #Selector implementation desn't exist 
```
```suggestion
    #Selector implementation doesn't exist 
```

m-fila commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+end
+
+description(sk::SoftKiller)::String = begin
+    #from definiton of is_initialised  in RectangularGrid.hh
```
```suggestion
    #from definitionof is_initialised in RectangularGrid.hh
```

graeme-a-stewart replied:
Also, what is `RectangularGrid.hh`? Something copied from Fastjet?

emadmtr replied:
Yes, RectangularGrid.hh is a header file is in FastJet and it defines a description() method to report information about the grid configuration but I replaced it with Julia's show as suggested.

Moelf commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+end
+
+tile_index(sk::SoftKiller,
+           p::PseudoJet)::Int64 = begin
```
why does this have to be Int64 not just `Int`

grasph replied:
why does the type need to be specified at all?

Moelf commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+        throw("SoftKiller not properly initialised.")
+    end
+
+    @assert all_tiles_equal_area() 
```
what's the point of this

graeme-a-stewart replied:
It also always returns `true`, so its not a real check of anything (TilingBase.jl:L26)...

m-fila commented on examples/softk.jl:
```diff
@@ -0,0 +1,281 @@
+                                          skipevents = args[:skip],
+                                          T = jet_type)
+
+    #Setting dimetions for Softkiller                                   
```
```suggestion
    #Setting dimensions for Softkiller                                   
```

m-fila commented on examples/softk.jl:
```diff
@@ -0,0 +1,281 @@
+                                                                       algorithm = algorithm)
+    @info "Jet reconstruction will use $(algorithm) with power $(p)"
+
+    #This is a vector of PseudoJets that contatins hard event and pileup 
```
```suggestion
    #This is a vector of PseudoJets that contains hard event and pileup 
```

m-fila commented on examples/softk.jl:
```diff
@@ -0,0 +1,281 @@
+    #This is a vector of PseudoJets that contatins hard event and pileup 
+    #this vector will get clustered without SoftKiller applied 
+    all_jets = PseudoJet[]
+    #This is a vector of PseudoJets that contatins hard event and pileup 
```
```suggestion
    #This is a vector of PseudoJets that contains hard event and pileup 
```

m-fila commented on examples/softk.jl:
```diff
@@ -0,0 +1,281 @@
+    origin= Dict{PseudoJet, String}()
+
+    Yi, Phii, pti, colorsr = Float64[], Float64[], Float64[], String[]
+    #Clustering and updating data for the gaphs plus filling all_jets_sk for the pile up 
```
```suggestion
    #Clustering and updating data for the graphs plus filling all_jets_sk for the pile up 
```

m-fila commented on examples/softk.jl:
```diff
@@ -0,0 +1,281 @@
+
+    pt_threshold = 0.00
+    soft_killer_event = PseudoJet[]
+    #Applying SoftKiller to a non-clistered vector of PseudoJets 
```
```suggestion
    #Applying SoftKiller to a non-clustered vector of PseudoJets 
```

m-fila commented on examples/softk_rt.jl:
```diff
@@ -0,0 +1,167 @@
+                                          skipevents = args[:skip],
+                                          T = jet_type)
+
+    #Setting dimetions for Softkiller                                   
```
```suggestion
    #Setting dimensions for Softkiller                                   
```

m-fila commented on examples/softk_rt.jl:
```diff
@@ -0,0 +1,167 @@
+                                                                      algorithm = algorithm)
+    @info "Jet reconstruction will use $(algorithm) with power $(p)"
+    
+    #This is a vector of PseudoJets that contatins hard event and pileup 
+    #but it will get clustered after SoftKiller was applied 
```
```suggestion
    #This is a vector of PseudoJets that contains hard event and pileup 
    #but it will get clustered after SoftKiller was applied 
```

m-fila commented on examples/softk_rt.jl:
```diff
@@ -0,0 +1,167 @@
+
+    pt_threshold = 0.00
+    soft_killer_event = PseudoJet[]
+    #Applying SoftKiller to a non-clistered vector of PseudoJets 
```
```suggestion
    #Applying SoftKiller to a non-clustered vector of PseudoJets 
```

m-fila reviewed (commented):
I spotted some typos. This will be a bit annoying but can we have space after `#` in comments?

m-fila commented on src/TilingBase.jl:
```diff
@@ -0,0 +1,47 @@
+#using JetReconstruction
+
+abstract type TilingBase end
```
I'm not sure whether this type is needed. It's only used to subtype `SoftKiller`, there are no functions for `TilingBase`

m-fila commented on src/TilingBase.jl:
```diff
@@ -0,0 +1,47 @@
+
+n_tiles()::Int64 = begin
+    return 0
+end
+
+n_good_tiles()::Int64 = begin
+    return n_tiles()
+end
+
+tile_is_good(itile::Int64)::Bool = begin
+    return true
+end
+
+all_tiles_good()::Bool = begin
+    return n_good_tiles() == n_tiles()
+end
+
+all_tiles_equal_area()::Bool = begin
+    return true
+end
+
+tile_area(itile::Int64)::Float64 = begin
+    return mean_tile_area()
+end
+
+mean_tile_area()::Float64 = begin
+    return 0
+end
+
+description()::String = begin
+    return 0
+end
+
+is_initialised()::String = begin
+    return 0
+end
+
+is_initialized()::Bool = begin
+    return is_initialised()
+end
```
```suggestion
```
These functions are never called! Also having both `is_initialized()` and `is_initialised()` would be very confusing

graeme-a-stewart replied:
Tangentially I consider that US English spellings are standard for code, so in method names it should be `initialized` and so on...

graeme-a-stewart commented:
> Hi @graeme-a-stewart, we weren't sure what to do with the formatting in the CI. When we ran `JuliaFormatter` it seemed to make changes that didn't cause that check to pass on my own fork, so we've left it undone for now.
> 
> We're anyway expecting to make some changes during review, so we thought this could be handled in parallel.

Hi

The CI checks the formatting, but it doesn't fix it for you (sorry, that's a TODO). The recipe to format correctly is, from the top of the package, run:

```sh
julia -e 'using JuliaFormatter; format(".")'
```

I have `JuliaFormatter` in my default package list (it's not part of the JetReco dependencies).

Indeed - it can be tweaked during the review process.

graeme-a-stewart commented:
Thank you @emadmtr - great to see this functionality being developed.

Let me make a few important generic comments that are not specific to lines of code:

- Can you rebase the PR to the head of `main`? There are quite a few important changes that are in the pipeline for preparing the next major release, and it's important to develop against these.
- There don't seem to be any CI tests. It's very important that each feature is tested, to make sure it keeps working. There are plenty of examples in the `test` directory of how to do this (see, e.g., `test-substructure.jl` - you should write a similar `test-softkiller.jl` and add it to `runtests.jl`). Usually, we use Fastjet to produce a reference output file - we have a helper function that allows these files to be compressed (`open_with_stream()` - and `zstd` is recommended now).
- Great that you have examples! We are factorising these into subdirectories of `examples`, so `examples/softkiller` would be the right place. I would rename the example scripts to be the full name `softkiller.jl`, just so that the purpose is crystal clear.
- One reason for doing examples like this is to factorise out dependencies for examples, particularly when these are heavy. Could you also move to using `Makie` instead of `Plots`? All of our other visualisations are done with `Makie`. This means have an example specific `Project.toml` file with your extra dependencies in it.
- Documentation: we need it! So you should add a `docs/src/softkiller.md` that explains how to use this feature. (And tweak `docs/make.jl` to compile that page.)

Thanks a lot!

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+    _ymin::Float64
+    _requested_drap::Float64
+    _requested_dphi::Float64
+    #selector 
```
This isn't a very informative comment (and space after the `#` please!)

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+    end
+
+    function SoftKiller(rapmax::Float64, grid_size::Float64)
+        print("2 variables\n")
```
Should be removed

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+        print("4 variables \n")
+        grid = new(rapmax, rapmin, drap, dphi, 0, 0, 0.0, 0.0, 0.0, 0.0, 0.0, 0, 0)
+        _setup_grid(grid)
+        print(description(grid))
```
Please don't print things for production code (in the worst case, if needed, this can be a `@debug`, but only if needed)

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+        grid = new(rapmax, -rapmax, grid_size, grid_size, 0, 0, 0.0, 0.0, 0.0, 0.0, 0.0, 0,
+                   0)
+        _setup_grid(grid)
+        print(description(grid))
```
Remove

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+    res + 1
+end
+
+function _setup_grid(sk::SoftKiller)
```
This is a mutating function - should be `_setup_grid!()`

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+    sk._cell_area = sk._dy * sk._dphi
+end
+
+description(sk::SoftKiller)::String = begin
```
If you do need this functionality then write it as a `show(io::IO, sk::SoftKiller)` function that outputs a string description of `sk`. See, e.g., https://github.com/JuliaHEP/JetReconstruction.jl/blob/b2926f5f38f81d2ef8929de6c4da8ed2e70553dc/src/Pseudojet.jl#L121.

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+    descr
+end
+
+n_tiles(sk::SoftKiller)::Int64 = begin
```
For `SoftKiller` is this useful? Usually you only need to write these accessor type functions if you are offering exposed functionality outside the package, or you need to do something special based on sub-types. It doesn't look like either of these is the case and in your code it's fine just to use `sk._ntotal` directly.

Please review other accessors using the same criteria!

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+    #fills the lector of length n_tiles with 0's
+    max_pt2 = fill(0.0, n_tiles(sk))
+
+    #starts from 1 not 0! 
```
Not a useful comment, I think!

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,169 @@
+    return filtered_events
+end
+
+function apply(sk::SoftKiller, event::Vector{PseudoJet}, reduced_event::Vector{PseudoJet},
```
It looks like this function mutates `reduced_event` so it should be `apply!`.

It also looks like `reduced_event` is completely rewritten - none of its original content it accessed and it's even resized. So this function signature looks weird to me. Why not just build `reduced_event` internally and then return it?

Moelf commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,144 @@
+        grid = new(rapmax, -rapmax, grid_size, grid_size, 0, 0, 0.0, 0.0, 0.0, 0.0, 0.0, 0,
+                   0)
+        _setup_grid(grid)
+        println(grid)  
```
```suggestion
```

Moelf commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,144 @@
+        end
+    end
+
+    resize!(reduced_event, length(indices))
+    for (i, idx) in enumerate(indices)
+        reduced_event[i] = event[idx]
+    end
```
```suggestion
    reduced_event = event[indices]
```

Moelf commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,144 @@
+        throw("SoftKiller not properly initialized.")
+    end
+
+    reduced_event = PseudoJet[]
```
```suggestion
```

emadmtr commented:
Hello, thank you everyone for taking the time to review my code and for the comments you left. Hopefully, I've addressed them appropriately, but please let me know if I've misinterpreted something and it needs more work.

The only comments that I am currently working on are making the test file and the documentation. I have a question regarding the documentation. Would you prefer if I were to use the original method description from the FastJet contrib file? As I believe the developers provided a very detailed explanation and I will include that the documentation is more or less cited from there with small changes to match the Julia implementation. 

Also, you mentioned that it is a good idea to add the sample files, however the sample files Professor LeBlanc and I have are pretty big, especially the pileup file. For now we’ve left them out, but we can definitely try some alternatives, such as reducing the pileup file, that wouldn't overwhelm the whole repo if that would be useful.

emadmtr commented:
Hello, I found some other spelling mistakes I've and also added something to the process_event in softkiller_plots to account for the recent changes in ClusterSequence.jl as my old implementation wasn't passing the newly added assertion for the clustering history. I also saw that I missed 1-2 comments as I was resolving them,so now I hope I addressed everything. I also pulled the test file so my changes shouldn't interfere with whatever Professor LeBlanc has submitted so far.

Moelf commented on .gitignore:
```diff
@@ -36,3 +36,6 @@ benchmark/*
 /debug/*
+
+# Input files for testing
+test/data/sk_example*.hepmc
```
Good question... I see a few options:

- if it's not too large (< 10 MB), just include it
- if it's medium large, you can use Git LFS maybe
- if it's even larger than that, and we have apetite for it, we can upload the file somewhere (like manually in a GitHub release), and use Julia's Artifact https://pkgdocs.julialang.org/v1/artifacts/#Basic-Usage

graeme-a-stewart replied:
Hello @mattleblanc - thanks for the comment. I would be inclined, for simplicity, to try to include test data files here. However, it's clear that we do not need 1000 events for testing - honestly 10 would be enough and then I would expect that, compressed, the hepmc file would be O(1MB), which is fine.

graeme-a-stewart commented on test/test-softkiller.jl:
```diff
@@ -0,0 +1,83 @@
+
+include("common.jl")
+
+using Random
```
I just did this so that the tests would at least run clean

graeme-a-stewart commented:
Hello @emadmtr - thanks a lot for the changes. I had a bit of trouble between your changes and @mattleblanc, so I fixed a few small things and also ran the formatter. More comments coming now...

mattleblanc commented:
Hi @graeme-a-stewart, we added a little initial documentation this afternoon that @emadmtr will continue (preparing the package-level documentation and adding comments to the examples, etc.). So what's here is not complete at the moment.

When trying to compile the documentation following the instructions in the package, the build seems to fail due to `EEJet` in `JetReconstruction.jl/ext/EDM4hepJets.jl`. 

On our other open point, I will push a change to include the input file for the examples soon. I'll try to use LFS for this, as the file is ~100MB.

Moelf commented:
we don't have PR documentation preview because this is PR is from a fork repo, but looks like documentation compiled just fine -- what did you try and failed?

emadmtr commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,133 @@
+    return filtered_events
+end
+
+function softkiller!(sk::SoftKiller, event::Vector{PseudoJet})
```
We added comments explaining what it does in SoftKiller.jl but if that's not what you mean please let us know.

emadmtr commented:
We added the test files and the documentation, everything seems okay on our side but if there is anything that we've missed let us know.

graeme-a-stewart commented on examples/softkiller/README.md:
````diff
@@ -0,0 +1,24 @@
+This example can be run as:
+
+```
+julia --project=examples examples/softkiller_plots.jl  --maxevents=100 --grid-size=0.4  --algorithm=Kt/ --pileup-file=test/data/sk_example_PU.hepmc --hard-file=test/data/sk_example_HS.hepmc 
````
Nice! However, can I request that the example be given assuming that this is the user's current working directory (it's the model we use in all of the other cases). So it should be more like...

```sh
julia --project softkiller_plots.jl  --maxevents=100 --grid-size=0.4  --algorithm=Kt --pileup-file=../../test/data/sk_example_PU.hepmc.zst --hard-file=test/data/sk_example_HS.hepmc.zst
```

- The name of the algorithm is also fixed (no trailing slash)
- The input files should be Zstd compressed

There is also a dependency issue here - the `examples` project does not have the correct dependencies for the plot example. So you need to generate a separate `Project.toml` file in this directory with all of the dependencies for these examples.

graeme-a-stewart commented on examples/softkiller/README.md:
````diff
@@ -0,0 +1,24 @@
+This example can be run as:
+
+```
+julia --project=examples examples/softkiller_runtime.jl  --maxevents=100 --grid-size=0.4  --algorithm=Kt/ --pileup-file=test/data/sk_example_PU.hepmc --hard-file=test/data/sk_example_HS.hepmc 
````
Comments as above

graeme-a-stewart commented on examples/softkiller/softkiller_plots.jl:
```diff
@@ -0,0 +1,277 @@
+using ArgParse
+using Profile
```
I don't think `Profile` is needed

graeme-a-stewart commented on examples/softkiller/softkiller_plots.jl:
```diff
@@ -0,0 +1,277 @@
+    else
+        jet_type = PseudoJet
+    end
+    @assert jet_type==PseudoJet "SoftKiller only supports PseudoJet"
```
This is a bit convoluted - it would make more sense just to do a positive check for `JetReconstruction.is_pp`.

graeme-a-stewart commented on examples/softkiller/softkiller_runtime.jl:
```diff
@@ -0,0 +1,136 @@
+using ArgParse
+using Profile
```
Again, I don't think `Profile` is useful

graeme-a-stewart commented on examples/softkiller/softkiller_runtime.jl:
```diff
@@ -0,0 +1,136 @@
+    else
+        jet_type = PseudoJet
+    end
+    @assert jet_type==PseudoJet "SoftKiller only supports PseudoJet"
```
Same comment as above - just test `is_pp`

graeme-a-stewart commented on examples/softkiller/softkiller_runtime.jl:
```diff
@@ -0,0 +1,136 @@
+
+    # Apply SoftKiller to all_jets_sk (hard + pileup) and 
+    # measure the runtime of the softkiller method 
+    @time softkiller!(soft_killer, all_jets_sk)
```
There is no sense in using a `@time` macro here

graeme-a-stewart commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,203 @@
+using JetReconstruction
+
+"""
+SoftKiller
```
This should be indented 4 spaces and be the struct signature:

```jl
   mutable struct SoftKiller
```

The more descriptive parts of this docstring about FastJet should be in the package documentation.

graeme-a-stewart reviewed (changes requested):
Great work @emadmtr and @mattleblanc - thanks for this. From my side we now just need to tidy up a few things:

- Example documentation should be rooted in the `examples/softkiller` directory and a `Project.toml` is needed
- The sample files should be zstd compressed, which saves a lot of space. Try, e.g., `zstd -19 sk_example_PU.hepmc`
- We need some package lavel documentation in, say, `docs/src/softkiller.md`

graeme-a-stewart commented on examples/softkiller/softkiller_plots.jl:
```diff
@@ -0,0 +1,277 @@
+           ],
+           ["Pileup", "Hard Event", "Jet"], "Legend")
+
+    save_dir = "examples/softkiller"
```
Assume that we're running in `examples/softkiller`, so you don't need this path.

Also, make sure there's a `.gitignore` to not accidentally commit the output PNG files.

m-fila commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,203 @@
+    _ny::Int64
+    _nphi::Int64
+
+    """
+        SoftKiller(rapmax::Float64, grid_size::Float64)
+
+    Construct a square SoftKiller grid from `-rapmax` to `rapmax` in rapidity, with tile size `grid_size`.
+    """
+    function SoftKiller(rapmin::Float64, rapmax::Float64, drap::Float64, dphi::Float64)
+        grid = new(rapmax, rapmin, drap, dphi, 0, 0, 0.0, 0.0, 0.0, 0.0, 0.0, 0, 0)
+        _setup_grid!(grid)
+        grid
+    end
+
+    """
+        SoftKiller(rapmin::Float64, rapmax::Float64, drap::Float64, dphi::Float64)
+
+    Construct a SoftKiller grid from `rapmin` to `rapmax` in rapidity, with tile sizes `drap` and `dphi`.
+    """
+    function SoftKiller(rapmax::Float64, grid_size::Float64)
+        grid = new(rapmax, -rapmax, grid_size, grid_size, 0, 0, 0.0, 0.0, 0.0, 0.0, 0.0, 0,
+                   0)
+        _setup_grid!(grid)
+        grid
+    end
+end
```
I think these two docstrings should be switched with each other

emadmtr commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,203 @@
+    _requested_drap::Float64
+    _requested_dphi::Float64
+    _ntotal::Int64
+    _ngood::Int64
```
I took it out!

emadmtr commented:
Hi @m-fila I made the changes you requested about the documentation and the _ngood parameter. About the user interface, we were following the C++ implementation but it definately sounds like something that we could revisit in a future PR. Also I tried running the Julia Formatting with the most recent instructions but it seems to fail, do you have anyadvice how to fix it?

m-fila commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,203 @@
+  Construct a square grid from `-rapmax` to `rapmax` in rapidity, with tile size `grid_size`.
+
+"""
+mutable struct SoftKiller
```
If I understood correctly the only function mutating `SoftKiller` is `_setup_grid!` which is only used in the constructors, so with some reorganization it should be possible to make it immutable. I think we might fix that later as well

emadmtr replied:
That sounds great, _setup_grid! was originally a method in RectangularGrid class however that was it's only functionality so we moved it to Softkiller. We can definately work on it to make it immutable later on!

graeme-a-stewart replied:
We can just call the grid setup *before* we construct the actual SK object, and then pass in the derived parameters. Then Softkiller can be immutable.

m-fila commented on src/SoftKiller.jl:
```diff
@@ -0,0 +1,203 @@
+    _inverse_dphi::Float64
+    _ny::Int64
+    _nphi::Int64
+
+    
+    """
```
```suggestion

    """
```

There is an extra line with whitespaces here which causes formatter check failure

m-fila reviewed (commented):
Thanks. I'm afraid I don't have any ideas how to help with formatter without looking at your setup.  But I marked the offending line in the suggestions

m-fila commented:
I think you could also add the softkiller documentation page in the `docs/make.jl` so it appears in the index in the generated documentation and gains more visibility

https://github.com/emadmtr/JetReconstruction.jl/blob/eae1d717562471bcd0444bf6b825b704d20bf76d/docs/make.jl#L11-L28

emadmtr commented:
Thank you for the quick follow up I just made the new changes, sorry for missing them. Before pushing again do you have any suggestions what I did that fails the CI check?

m-fila commented:
Sorry I pushed before reading you message. Please git fetch and git pull to synchronize. 

There were two problems
- tests still calling `softkiller!` but the function was renamed
- In the automatic tests we are checking with both latest version of Julia (1.11) and long-term-support (LTS) (1.10). The `Statistics` is a standard library and has the same version of Julia used. In the Project.toml a constrain was put that `Statistics` must be 1.11 or newer, but with LTS we could only get 1.10 so we couldn't fulfill the requirements

m-fila reviewed (approved):
Thank you for your patience. I think that's all from my side
LGTM!

graeme-a-stewart commented:
@all-contributors please add @m-fila for review
